File: app/src/main/java/dev/alenajam/opendialer/adapter/CallOptionsAdapter.java
package dev.alenajam.opendialer.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import dev.alenajam.opendialer.R;
import dev.alenajam.opendialer.model.CallOption;

import java.util.ArrayList;

public class CallOptionsAdapter extends RecyclerView.Adapter<CallOptionsAdapter.ViewHolder> {

  private ArrayList<CallOption> options;
  private Context context;
  private CallOptionsAdapterListener listener;

  public CallOptionsAdapter(ArrayList<CallOption> options, Context context, CallOptionsAdapterListener listener) {
    this.options = options;
    this.context = context;
    this.listener = listener;
  }

  @NonNull
  @Override
  public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    Context context = parent.getContext();
    LayoutInflater inflater = LayoutInflater.from(context);

    // Inflate the custom layout
    View contactView = inflater.inflate(R.layout.item_call_option, parent, false);

    // Return a new holder instance
    ViewHolder viewHolder = new ViewHolder(contactView);
    return viewHolder;
  }

  @Override
  public void onBindViewHolder(@NonNull ViewHolder holder, final int position) {
    CallOption currentOption = options.get(position);

    holder.title.setText(currentOption.getText());
    holder.icon.setImageDrawable(context.getDrawable(currentOption.getIcon()));
    holder.layout.setOnClickListener(v -> listener.onClick(currentOption.getId()));
  }

  @Override
  public int getItemCount() {
    return options.size();
  }

  public void addItem(CallOption item) {
    this.options.add(item);
    notifyItemInserted(this.options.size() - 1);
  }

  public int getItemIndex(CallOption option) {
    return this.options.indexOf(option);
  }

  @Nullable
  public CallOption getItem(int id) {
    for (int i = 0; i < this.options.size(); i++) {
      if (this.options.get(i).getId() == id) {
        return this.options.get(i);
      }
    }
    return null;
  }

  public interface CallOptionsAdapterListener {
    void onClick(int whichOption);
  }

  // Provide a direct reference to each of the views within a data item
  // Used to cache the views within the item layout for fast access
  protected class ViewHolder extends RecyclerView.ViewHolder {
    // Your holder should contain a member variable
    // for any view that will be set as you render a row
    private TextView title;
    private ImageView icon;
    private LinearLayout layout;

    // We also create a constructor that accepts the entire item row
    // and does the view lookups to find each subview
    private ViewHolder(View itemView) {
      // Stores the itemView in a public final member variable that can be used
      // to access the context from any ViewHolder instance.
      super(itemView);
      title = itemView.findViewById(R.id.title);
      icon = itemView.findViewById(R.id.icon);
      layout = itemView.findViewById(R.id.layout);
    }
  }
}

File: feature/calls/src/main/java/dev/alenajam/opendialer/feature/calls/CallOptionsAdapter.kt
package dev.alenajam.opendialer.feature.calls

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import dev.alenajam.opendialer.data.calls.CallOption

class CallOptionsAdapter(
  private var options: List<CallOption> = emptyList(),
  private val onClick: (option: CallOption) -> Unit
) : RecyclerView.Adapter<CallOptionsAdapter.ViewHolder>() {
  inner class ViewHolder(inflater: LayoutInflater, private val parent: ViewGroup) :
    RecyclerView.ViewHolder(inflater.inflate(R.layout.item_call_log_child, parent, false)) {
    private val layout = itemView.findViewById<LinearLayout>(R.id.layout)
    private val icon = itemView.findViewById<ImageView>(R.id.icon)
    private val text = itemView.findViewById<TextView>(R.id.text)

    fun bind(option: CallOption) {
      //icon.setImageDrawable(ContextCompat.getDrawable(parent.context, option.icon))

      val resId = when (option.id) {
        CallOption.ID_CREATE_CONTACT -> R.string.create_new_contact
        CallOption.ID_ADD_EXISTING -> R.string.add_to_a_contact
        CallOption.ID_SEND_MESSAGE -> R.string.send_message
        CallOption.ID_CALL_DETAILS -> R.string.call_details
        CallOption.ID_COPY_NUMBER -> R.string.copy_number
        CallOption.ID_EDIT_BEFORE_CALL -> R.string.edit_number_before_call
        CallOption.ID_BLOCK_CALLER -> R.string.blockThisCaller
        CallOption.ID_UNBLOCK_CALLER -> R.string.unblockThisCaller
        CallOption.ID_DELETE -> R.string.delete
        else -> null
      }

      resId?.let {
        text.text = parent.context.getString(resId)
      }

      layout.setOnClickListener { onClick(option) }
    }
  }

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
    ViewHolder(LayoutInflater.from(parent.context), parent)

  override fun getItemCount(): Int = options.size

  override fun onBindViewHolder(holder: ViewHolder, position: Int) =
    holder.bind(options[position])
}
